### PR TITLE
Reimplement ModelLoaderRegistry and ICustomModelLoader

### PR DESCRIPTION
**Pull request by tterrag1098**
This is mainly just for code review. I have a bit of cleanup left to do, but I want to give people an idea of the upcoming changes as well as allow feedback on what's been done.

### CODE REVIEW

File: patches/minecraft/net/minecraft/client/renderer/model/ModelManager.java.patch [hunk 1]
```diff
@@ -5,7 +5,7 @@
  
     public void func_195410_a(IResourceManager p_195410_1_) {
 -      ModelBakery modelbakery = new ModelBakery(p_195410_1_, this.field_174956_b);
-+      net.minecraftforge.client.model.ModelLoader modelbakery = new net.minecraftforge.client.model.ModelLoader(p_195410_1_, this.field_174956_b);
++      ModelBakery modelbakery = new ModelBakery(p_195410_1_, field_174956_b);
```
Comment (Chocohead) [inline patches/minecraft/net/minecraft/client/renderer/model/ModelManager.java.patch]: Could this patch not be dropped completely if `this` was added back?

Comment (tterrag1098) [reply patches/minecraft/net/minecraft/client/renderer/model/ModelManager.java.patch]: @Chocohead yes, good catch

**Comment by williewillus** [comment]
what's the plan regarding MRL's only being allowed to point to blocks that exist, or item jsons?
as in, each MRL is `namespace:path#variant`, and there's new logic in ModelBakery that checks if `namespace:path` is an actual block.

**Comment by williewillus** [comment]
Could #5667 be implemented here? I'm thinking just a simple list of `ResourceLocation` in `ModelRegistryEvent`, MRL's in the list pointing to blockstate jsons and RL's pointing directly to model paths.

**Comment by tterrag1098** [comment]
@williewillus This is more focused on hooking up the ICustomModelLoader system, and not the bake event. That is an issue that needs fixing, but at this point it would be better to wait for 1.14 to deal with that, and get this in now before we start porting.

**Comment by Muramasa-** [comment]
Will this make it in 1.13, or is it being pushed to 1.14?

**Comment by mura3277** [comment]
@tterrag1098 May I ask, will we be able to redirect item/block models to somewhere other than their default location? This would be very useful and convenient for organisation (putting type of jsons in sub folders) and to allow using json templates.  Fabric has a nice api for things like this ([ModelVariantProvider](https://grondag.github.io/renderosity/fabric-api-javadoc/net/fabricmc/fabric/api/client/model/ModelVariantProvider.html) & [ModelResourceProvider](https://grondag.github.io/renderosity/fabric-api-javadoc/net/fabricmc/fabric/api/client/model/ModelResourceProvider.html))

Alternatively, a similar system to item "blockstates" would also be very useful.

**Comment by Vazkii** [comment]
Any update on this? It's pretty much essential for multiple of my mods to work on 1.14.

**Comment by henkelmax** [comment]
For me too. 
For now I made a workaround with item overrides.

**Comment by tterrag1098** [comment]
This is (obviously) still going to happen but this approach is dead as it had some critical flaws and would be far to much effort to port to 1.14. So I'll be starting from scratch there when I find the time.

**Pull request closed without merging**